src/chocolatey/infrastructure.app/nuget/ChocolateyNugetCredentialProvider.cs

```diff
@@ -86,67 +86,62 @@ public Task<CredentialResponse> GetAsync(Uri uri, IWebProxy proxy, CredentialReq
             }
 
             // credentials were not explicit
-            // discover based on closest match in sources
-            var candidateSources = _config.MachineSources.Where(
-                s =>
-                {
-                    var sourceUrl = s.Key.TrimEnd('/');
-
-                    try
-                    {
-                        var sourceUri = new Uri(sourceUrl);
-                        return sourceUri.Host.IsEqualTo(uri.Host)
-                            && !string.IsNullOrWhiteSpace(s.Username)
-                            && !string.IsNullOrWhiteSpace(s.EncryptedPassword);
-                    }
-                    catch (Exception)
-                    {
-                        this.Log().Error("Source '{0}' is not a valid Uri".FormatWith(sourceUrl));
-                    }
-
-                    return false;
-                }).ToList();
-
+            // find matching source(s) in sources list
+            var trimmedTargetUri = new Uri(uri.AbsoluteUri.TrimEnd('/'));
             MachineSourceConfiguration source = null;
 
-
-            if (candidateSources.Count == 1)
+            // If the user has specified --source with a *named* source and not a URL, try to find the matching one
+            // with the correct URL for this credential request.
+            var namedExplicitSources = _config.ExplicitSources?.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries)
+                .Where(s => !Uri.IsWellFormedUriString(s, UriKind.Absolute))
+                .ToList();
+            if (namedExplicitSources?.Count > 0)
             {
-                // only one match, use it
-                source = candidateSources.FirstOrDefault();
+                // Uri.Equals() and == operator compare hostnames case-insensitively and the remainder of the url case-sensitively
+                // while ignoring #fragments on the URLs, but does care about trailing slashes, which we do not here.
+                source = _config.MachineSources
+                    .Where(s => namedExplicitSources.Contains(s.Name) && new Uri(s.Key.TrimEnd('/')) == trimmedTargetUri)
+                    .FirstOrDefault();

```

Comment:
Why are we doing case sensitive checking here?

The `Contains` method is case sensitive, and as the comment says it does a case sensitive match on the url (just not the hostname).
That do not make sense, and I would think it is a case insensitive match we would want.

Reply:
The reason for doing a case sensitive match on everything but the hostname is that the hostname is handled by DNS and is case insensitive, but the rest of the URL is up to the responding server which may or may not be case insensitive. For example: `https://docs.chocolatey.org/en-us/central-management/` is valid while `https://docs.chocolatey.org/en-us/central-managemenT/` is not. But if you were to case-insensitive compare them they are identical.

Reply:
While that is true for the documentation site we have, that is actually quite uncommon.

If we take the api we have, the URL `https://community.chocolatey.org/api/v2/` returns the exact same information as `https://community.chocolatey.org/API/v2/`.

But even if they were not returning the same information, it would not be the point.

The point is, we are looking up information about a URL in the machine sources configuration, and it is not guaranteed that a user will remember the exact case of the URI that was configured but still intend to use it. By not allowing the case to be different, we are preventing this configured URL to be picked up. We should also use the URL as specified in the machine configuration (and if I am reading everything correctly, we are doing that), so even if the server would accept one case but not the other, it wouldn't be a problem since it already uses the case that is configured on the machine object.

There is also the case that with this change, a user can not use a named source that differs in casing either. Ie, they can not pass the source `Chocolatey` when the name is configured to be `chocolatey`. This is something that is changing the behavior of what is currently available, as currently using a named source with a different case will resolve to a configured URL.

Reply:
I think I have talked myself into what you're suggesting here. On the topic of the URL case sensitivity, I'll say this: in my experiences, casing differences in URLs outside of the domain name are typically up to the host operating system. I also know that Nexus is partially case sensitive for instance `http://nexus/repository/my-repo/` and `http://nexus/repository/My-RePO/` are both treated the same, but `http://nexus/Repository/my-repo/` results in a 404 error.

But getting to the point of allowing the user to put the source in at the terminal and not have to worry about the casing of it, I think I can get behind that for you reason you said about we're pulling the URL from the machine config, so even if they got the casing wrong, we're still going to use the correct casing.

The named source is still being picked up in a case insensitive matter (at least in my limited testing it was).

Reply:
Small wrench to throw in Kim's logic here: this credential provider _does not_ provide the URL back to NuGet. NuGet is still potentially going to be querying an incorrectly-cased URL with the saved credentials, because we don't do this kind of lookup anywhere else when evaluating which sources to query. So this is a narrower case but still potentially the same as the original bug: querying route A with credentials intended for route B on the same server.

That said, given this is still an improvement and it would take a long time to hash that out I'm content to leave that to a follow up issue / let it be until we remove this weird lookup logic entirely (I'll probably put some details around things Gary and I have talked about in the related issue @corbob raised and we can figure out how we want to split things out from there, if needed).

Reply:
I have updated this to do a case sensitive comparison of the lower case name (because we can't tell `.Contains` to do a case insensitive compare, and using `string.Join` to join into a string leaves potential for a substring matching where it shouldn't). And then to do a case insensitive compare on the URI Components that the `==` compares on.